### PR TITLE
Begin consolidating repetitive chart code and add "Case Terminations by Termination Type" chart to the ND revocations page

### LINKED ISSUE #187: Add front end support for "case terminations by month" chart
**Is your feature request related to a problem? Please describe.**
North Dakota currently uses case based terminations for calculation of revocation and success rates. This chart will include more of the numbers ND is used to in order to balance their current calculations with the methodology (ours) they want to move towards. 

**Describe the solution you'd like**
A new chart, on the "revocations" page, that examines MoM supervision termination by type, a la the `REVOCATIONS BY VIOLATION TYPE` chart. This would use our supervision termination type enums (absconsion, death, discharge, expiration, external unknown, revocation, suspension). This would be an event based chart, not a person based chart. Naively, it would use only data ingested from docstars and the methodology would allow ND users to understand how a case in our graph would have arrived there from their system. 

Initial thinking is to have this chart be placed under the `Revocations by supervision type` chart. It would not have a map mode, but would support rates & counts, different time periods, parole vs. probation, and P&P district selection. Rate mode would be similar to the revocations by violation type rate mode, in which one can find out the percent of case terminations in each month of a particular type. 

**Describe alternatives you've considered**
- A revocations chart that uses the current ND methodology, which counts case terminations marked as revocation in Docstarts / case terminations in docstars. However, calculating revocations multiple ways seems non-intuitive from a UX perspective.

### PR DESCRIPTION
**Pull request by itsjon**
## Description of the change

[NOTE: This change is *far far smaller* than it seems! Those 100k+ new lines are snapshots to help me be sure I'm not breaking anything.]

This PR consolidates some repetitive code between the `RevocationCountBySupervisionType.js` and `RevocationCountByViolationType.js` charts and reuses the consolidated code to add the new `CaseTerminationsByTerminationType.js` chart.

To provide high confidence this PR doesn’t change the existing behavior, this PR contains multiple commits:

1. The first commit:
  - does basic refactoring to make the code that generates the two preexisting charts more testable,
  - snapshots all the states of each chart that are possible using the available filters and the existing fixture data, and
  - adds tests to verify the code generates those states
2. The second commit begins consolidation of duplicate code between `RevocationCountBySupervisionType.js` and `RevocationCountByViolationType.js`
3. The third commit adds the new "Case Terminations by Termination Type" chart

The tests using the snapshots pass at each commit (`yarn test`).

This isn't everything I'd wanted it to be in terms of refactoring, but I think it's a few steps in the right direction.

@kbinney the fixture data for this new chart at the moment contains all the termination types you listed in the Google Doc except "Unknown". Let's discuss whether or not that's a blocker to shipping this, or if we can add that subsequently. I'll also note that the bars are pretty short compared to the mockup, so let me know if that's a blocker.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

Closes #187

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Lint rules pass locally
- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Comment by itsjon** [comment]
@jessex sorry, I realize this is pretty innavigable in GitHub. Obviously this hundreds-of-snapshots approach to testing won't be the default approach, and I plan to replace these tests with simpler and faster unit tests once I've more fully consolidated the code. If you open the files list, once you hit the long list of files with `/snapshots/` in the path, scroll to the bottom to find the other actual code (`snapshotChartVariants.js` and on).

File: src/views/tenants/us_nd/Revocations.js [hunk 1]
```diff
@@ -422,6 +424,75 @@ const Revocations = () => {
...
+                  </div>
+                  <div className="collapse" id="collapseMethodologyCaseTerminationsByTerminationType" aria-labelledby="methodologyHeadingCaseTerminationsByTerminationType" data-parent="#methodologyCaseTerminationsByTerminationType">
+                    <div>
+                      <ul>
+                        <li>
+                          This chart includes counts based on case, not person. If a person on supervision has multiple
+                          cases, each case termination will be counted in the chart.
+                        </li>
+                        <li>
+                          Case terminations are included based on termination date in DocStars.
```
Comment (jessex) [inline src/views/tenants/us_nd/Revocations.js]: Nit: DocStars --> Docstars. Here and elsewhere.

**Comment by kbinney** [comment]
Looks good to me!

The docstars spelling is on me - sorry should be how josh put it: Docstars. I also realized the methodology is missing information on the unknown field. Can you add to the end of the comment on mapping termination types to termination code "Other is cases terminated with code 14"?

The actual data does have the unknown field, which we are actually going to call "other." I'll defer to @jessex if that change (adding to code and/or fixture) is a blocker.

**Comment by jessex** [comment]
Yes, @itsjon, we did finally conclude the full set of termination types to chart, and there will be one more field in the two fixture file with a key of "other". I will try to push up that change in a new commit direct to your branch shortly...

**Comment by jessex** [comment]
@itsjon I just pushed a commit to your branch that adds an _other_ field to each of the case termination chart fixture files. So now you should include that as one of the stacks in the bar chart.

**Comment by itsjon** [comment]
Thank you, @jessex and @kbinney! I'll make all these updates ASAP. You're correct @jessex that the tests will start failing in April--great catch! I though I could fix that quickly with a mock, but I hit an unexpected issue I'm working through.

File: src/components/charts/revocations/RevocationCountByViolationType.test.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+describe('getBarChartDefinition', () => {
+  const data = readJsonLinesFile(
+      path.join(__dirname, 'test_data/RevocationCountByViolationType/revocations_by_violation_type_by_month.json')
+  );
+
+  test('produces the expected chart definitions for all snapshotted scenarios', () => {
+    fs.readdirSync(path.join(__dirname, 'test_data/RevocationCountByViolationType/snapshots')).forEach(fileName => {
+      try {
+        const filters = fileName.slice(0, 0 - '.json'.length).split('|').reduce((filters, part) => {
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/RevocationCountByViolationType.test.js]: Yes, I'd had in mind to consolidate the tests when I further consolidate the components, but I've gone ahead and extracted that duplicate code.

File: src/components/charts/revocations/RevocationCountByViolationType.test.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+          return filters;
+        }, {});
+
+        const expectedDefinition = JSON.parse(fs.readFileSync(
+            path.join(__dirname, 'test_data/RevocationCountByViolationType/snapshots', fileName),
+            'utf8'
+        ));
+
+        let definition = getBarChartDefinition(Object.assign(filters, {revocationCountsByMonthByViolationType: data}));
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/RevocationCountByViolationType.test.js]: Fixed!

File: src/components/charts/revocations/RevocationCountByViolationType.test.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        const filters = fileName.slice(0, 0 - '.json'.length).split('|').reduce((filters, part) => {
+          const [name, value] = part.split(':');
+
+          filters[name] = value;
+
+          return filters;
+        }, {});
+
+        const expectedDefinition = JSON.parse(fs.readFileSync(
+            path.join(__dirname, 'test_data/RevocationCountByViolationType/snapshots', fileName),
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/RevocationCountByViolationType.test.js]: It is a bit weird! I used the code in `snapshotChartVariants.js`. We could maybe keep these snapshots around in order to flag any unexpected changes in the charts, but they are unwieldy, and I think I'll replace the snapshot-based tests with unit tests once more of the duplicate code is consolidated. While I'm in that process though, having an exhaustive set of snapshots gives me a lot more confidence I'm not changing any behavior.

File: src/components/charts/revocations/test_data/snapshotChartVariants.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    chartDefinitionFn: revocationCountByViolationType.getBarChartDefinition,
+    sampleData: readJsonLinesFile(
+        path.join(__dirname, 'RevocationCountByViolationType/revocations_by_violation_type_by_month.json')
+    ),
+    dataArg: 'revocationCountsByMonthByViolationType'
+  }
+];
+
+const filterOptions = {
+  metricType: ['counts', 'rates'],
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/test_data/snapshotChartVariants.js]: Yeah, I've been thinking about that. I'm not sure the snapshot tests will stick around for long, so I'm not sure that's worth it. How about we leave this as-is for now, and I write up a GH issue describing what I have in mind for the more comprehensive version of this refactor, for discussion?

File: src/components/charts/revocations/test_data/snapshotChartVariants.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
+/* Ad-hoc script for creating snapshots of charts with every available filter combination. */
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/test_data/snapshotChartVariants.js]: Done

File: src/components/charts/revocations/Shared.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+export const getChartDefinition = ({
+    chartId,
+    countsByMonth,
+    metricType,
+    numMonths,
+    filters,
+    bars,
+    barColorPalette
+}) => {
+  // TODO(jon): Try to streamline this function more.
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/Shared.js]: Done (https://github.com/Recidiviz/pulse-dashboard/issues/233)

File: src/components/charts/revocations/RevocationCountByViolationType.js [hunk 8]
```diff
@@ -18,159 +18,28 @@
...
+import { COLORS_FIVE_VALUES } from '../../../assets/scripts/constants/colors';
 import { configureDownloadButtons } from '../../../assets/scripts/utils/downloads';
-import {
-  toggleLabel, getMonthCountFromMetricPeriodMonthsToggle, updateTooltipForMetricType,
-  filterDatasetBySupervisionType, filterDatasetByDistrict,
-  toggleYAxisTicksStackedRateBasicCount,
-} from '../../../utils/charts/toggles';
-import { sortFilterAndSupplementMostRecentMonths } from '../../../utils/transforms/datasets';
-import { monthNamesWithYearsFromNumbers } from '../../../utils/transforms/months';
+import { getChartDefinition } from './Shared';
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/RevocationCountByViolationType.js]: Done

File: src/assets/scripts/constants/colors.js [hunk 9]
```diff
@@ -342,6 +342,16 @@ const COLORS_FIVE_VALUES_LIGHT = [
   '#fdb0a8',
 ];
 
+const COLORS_SEVEN_VALUES = [
```
[the remark this answers is not in the document]
Comment (itsjon) [reply src/assets/scripts/constants/colors.js]: I took them from the mocks @kbinney fleshed out for this chart

File: src/views/tenants/us_nd/Revocations.js (hunk 1 above)
Comment (itsjon) [reply src/views/tenants/us_nd/Revocations.js]: Fixed

File: src/views/tenants/us_nd/Revocations.js [hunk 10]
```diff
@@ -469,21 +469,22 @@ const Revocations = () => {
...
                           revocation counts, this chart <span className="font-weight-bold">does not</span> only examine
                           revocations resulting in admission to a DOCR facility.
                         </li>
                         <li>
                           Absconsion is all cases terminated with termination code 13. Revocation is all cases
                           terminated with code 9 or 10. Suspension is cases terminated with code 3 or 6. Discharge is
                           cases terminated with code 1, 2, 5, 8, 12, 15, 16, 17, or 18. Expiration is cases terminated
-                          with code 4, 7, 19, or 20. Death is cases terminated with code 11.
+                          with code 4, 7, 19, or 20. Death is cases terminated with code 11. Other is cases terminated
+                          ith code 14.
```
Comment (jessex) [inline src/views/tenants/us_nd/Revocations.js]: Whoops, typo: ith -> with

File: src/components/charts/revocations/RevocationCountByViolationType.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jessex) [reply src/components/charts/revocations/RevocationCountByViolationType.test.js]: @itsjon I'm not 100% sure. Has this been fixed? You have a longer comment below indicating that you ran into issues making this work.

File: src/components/charts/revocations/RevocationCountByViolationType.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jessex) [reply src/components/charts/revocations/RevocationCountByViolationType.test.js]: Very curious to discuss your experiences with snapshot testing! I want to understand enough to decide if we should go forward with that approach or not across all React components.

File: src/assets/scripts/constants/colors.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (jessex) [reply src/assets/scripts/constants/colors.js]: Nice!

**Comment by jessex** [comment]
@itsjon this is looking good! I just have one nit and one question about your tests, and there's still a merge conflict. We're just about done here.

File: src/views/tenants/us_nd/Revocations.js (hunk 10 above)
Comment (itsjon) [reply src/views/tenants/us_nd/Revocations.js]: Ah, silly. Fixed!

**Comment by itsjon** [comment]
OK, merge conflicts resolved and typo fixed. @jessex is the question about the tests the one you Slacked me about?

**Comment by jessex** [comment]
No, @itsjon, I was talking [about this one](https://github.com/Recidiviz/pulse-dashboard/pull/231#discussion_r391944743):

>@itsjon I'm not 100% sure. Has this been fixed? You have a longer comment below indicating that you ran into issues making this work.

File: src/components/charts/revocations/RevocationCountByViolationType.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (itsjon) [reply src/components/charts/revocations/RevocationCountByViolationType.test.js]: Sorry I missed this earlier. Yes, that's fixed via the mocked `Date` in `snapshotTesting.js`.

**Pull request merged**